### PR TITLE
Add Storybook to the project

### LINKED ISSUE #563: Storybook  usage
[Storybook](https://storybook.js.org/) is a nice tool that may help to sync Zeplin, ZeroHeight and more tools alltogether.
Definetely need to think about it.

### PR DESCRIPTION
**Pull request by julien-deramond**
Closes #563 

---

Preview: https://deploy-preview-565--boosted.netlify.app/storybook/

---

### High priority

- [x] Study Storybook sync with tools:
  - [x] Zeplin
  - [x] ZeroHeight
- [x] If it is not needed to have dynamically properties, replace knobs addon with controls addon (more limited than knobs but this is an official addon)
- [x] Add all the components
- [x] Advanced theming : process in progress to have a specific Orange Open Source project to share a common configuration
- [x] Stories automation:
  - Try to use our Hugo example shortcode (or any new one) to add stories content/examples?
- [ ] Add some guidance in the wiki: 
  - e.g. if automation ↑ is not possible, we'll need to add related entries in DoDs: 
    - for feature, add a new story; 
    - for release: check stories got updated accordingly to latest changes.
- ~~Mention Storybook somewhere in our docs? Is that useful at all?~~

### Medium priority

- [ ] Try the official addons
  - [ ] [Accessibility](https://github.com/storybookjs/storybook/tree/master/addons/a11y)
    - [x] Basic usage
    - [ ] Advanced usage
  - [ ] [Actions](https://storybook.js.org/docs/react/essentials/actions)
  - [ ] [Backgrounds](https://storybook.js.org/docs/react/essentials/backgrounds)
  - [ ] [Console](https://github.com/storybookjs/storybook-addon-console)
  - [x] [Controls](https://storybook.js.org/docs/react/essentials/controls).
  - [ ] [Docs](https://storybook.js.org/docs/react/writing-docs/introduction)
  - ~~[Knobs](https://github.com/storybookjs/storybook/tree/master/addons/knobs)~~
  - [ ] [Links](https://github.com/storybookjs/storybook/tree/master/addons/links)
  - [ ] [Source](https://github.com/storybookjs/storybook/tree/master/addons/storysource)
  - [ ] [Storyshots](https://github.com/storybookjs/storybook/tree/master/addons/storyshots)
  - [ ] [Toolbars](https://storybook.js.org/docs/react/essentials/toolbars-and-globals)
  - [x] [Viewport](https://storybook.js.org/docs/react/essentials/viewport)
- [ ] What about [community addons](https://storybook.js.org/addons)? Or writing an addon?
- [ ] Add the global design rules?
- [x] Try to apply some theming;
- [x] Be able to deploy branches to specific temporary URLs to show a WIP

## Integration

Default Storybook configuration created like this:
```
nvm use default # npm --version v6.14.5
npx sb init # chose HTML
npm i core-js --save-dev
```

## How to test

```
npm i
npm run storybook
```

---

### Chromatic

Was setup for deployment convenience, but replaced with Storybook static build + Hugo mounts.

- [x] Deploy Storybook v5-dev branch to a unique URL via GitHub Actions using Chromatic
  - Chromatic: https://www.chromatic.com/library?appId=5fc5eb7c9131880021f5e49c
  - Permalink: https://storybook--5fc5eb7c9131880021f5e49c.chromatic.com/
- [x] Link the Chromatic project to this GitHub project (https://www.chromatic.com/docs/access). Otherwise, users must sign in to access to Storybook

### CODE REVIEW

File: .storybook/preview-head.html
Comment (ffoodd) [inline .storybook/preview-head.html:7]: Should use v5 (without jQuery nor OrangeIcons). Can be a simple switch from `@4.5.3` to `@next` (for now) with the renaming of `orangeHelvetica` to `orange-helvetica`.

File: stories/1-Button.stories.js
Comment (ffoodd) [inline stories/1-Button.stories.js:1]: Tell me if we need to adapt ESlint configuration—or if we add a scoped one for stories maybe?

**Comment by ffoodd** [comment]
Thanks for starting this!

Some sidenotes for the future—we may:
- try to apply some theming
- try to use our Hugo example shortcode (or any new one) to add  stories content / examples

That'd be last things to do obviously, having a MVP is the main goal :)

File: .storybook/preview-head.html
Comment (julien-deramond) [reply .storybook/preview-head.html]: You're right, fixed it!

File: stories/1-Button.stories.js
Comment (julien-deramond) [reply stories/1-Button.stories.js]: Those were the type of problems that I fixed:

- `error  Filename is not in kebab case. Rename it to `alert.stories.js`
- `error Extra semicolon`
- `error Multiple spaces found before '='`

I don't think we need for the moment to adapt ESlint configuration.

**Comment by ffoodd** [comment]
It'd also need some love on the GitHub action part, since docs, pa11y and percy are required but won't be triggered outside of the docs! I'll find a way :)

**Comment by julien-deramond** [comment]
ℹ️ This comment will help me to track what I'm doing to try to connect Zeplin and Storybook.

- [ ] [Zeplin Connected Components Documentation](https://github.com/zeplin/connected-components-docs/blob/master/README.md#getting-started)
  - [ ] [getting started guide for all platforms/frameworks](https://github.com/zeplin/connected-components-docs/blob/master/docs/gettingStarted/ALL.md)
  - [ ] [Adding Storybook links](https://github.com/zeplin/connected-components-docs/blob/master/docs/link/STORYBOOK.md)
  - [ ] (Try it later) [Adding repository links, e.g. GitHub](https://github.com/zeplin/connected-components-docs/blob/master/docs/link/REPOSITORY.md)

**Comment by julien-deramond** [comment]
ℹ️ This comment will help me to track what I'm doing to try to connect ZeroHeight and the rest of the World.

- [x] Create an account on ZeroHeight

**Comment by ffoodd** [comment]
So after a few tries:
- Dropped Chromatic for now (keeping it in commit history for now, just in case → should be dropped completely through `rebase` if we don't go back)
- Use Storybook internal `build` command to get a static version, in `./site/storybook/`
- Use Hugo `mounts` feature to get it through Hugo generation, moving it to `/storybook` path.

So as of now, Chromatic isn't needed and might be ignored if its design-collaboration features don't get GDL attention.

Note for future self:
- Keeping accessibility tests in Storybook: useful or not? since we already check docs through Pa11y.
- Should `/storybook` be added to Percy snapshots? I'd say no since we alreaédy check guidelines, and have a monthly quota for snapshots.

**Comment by ffoodd** [comment]
If/when [IOT-Map-Component](https://github.com/Orange-OpenSource/IOT-Map-Component) gets linked to Boosted in any way, [Storybook composition](https://storybook.js.org/docs/react/workflows/storybook-composition) might be welcome.

**Comment by julien-deramond** [comment]
Just for info, I've modified the Zeroheight page used in this PR to embed Storybook components now deployed in Netflify (instead of Chromatic - dropped): https://zeroheight.com/7183633e4/p/18af52-alerts-netlify.

**Comment by julien-deramond** [comment]
Some modifications via https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/pull/565/commits/b1a8cb7393c23ca18d14f7ff399628c786c34dd9
- Bump @storybook/* from ^6.1.1 to ^6.2.9
- Get rid of @storybook/addons-knobs and replace it with Controls
- Add summary into the Controls area

File: .storybook/OrangeTheme.js
Comment (ffoodd) [inline .storybook/OrangeTheme.js:39]: Hell no, please use Boosted's dist logo instead. It's really, really better : 
1. Not another domain to get something from,
2. Responsive (includes the small logo) and optimized: see [Boosted v5 Design guidelines](https://v5-dev--boosted.netlify.app/docs/5.0/guidelines/logos/#responsive-svg).

Maybe send Boosted's logo to XD Labs x)

Comment (julien-deramond) [reply .storybook/OrangeTheme.js:39]: Yep. It was just a test asked by the designers.

**Comment by julien-deramond** [comment]
Automating the retrieval of components from Boosted documentation:
- [ ] 🐛 Rebase from beta3 to have working JavasScript
- [ ] 🐛 Find a way to remove the comment from the examples because it breaks the generated stories
- [ ] Remove `line-reader` from `package.json`.
- [ ] Remove `run-from-markdown.js` not used anymore.
- [ ] Deploy manually the built static version of Storybook in `gh_pages` under `/storybook/5.0/`
- [ ] Don't store the automatically generated stories in the repo but create them dynamically while generating the static version of Storybook
- [ ] Embed assets/resources (images, etc.) in the static version of Storybook

File: .storybook/OrangeTheme.js
Comment (julien-deramond) [reply .storybook/OrangeTheme.js]: Fixed with https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/pull/565/commits/eed9b92ea8e67fa98122558b1d26f1e580c2d087

File: .storybook/preview-head.html
Comment (ffoodd) [inline .storybook/preview-head.html:3]: I think you should be able to use sole shortcut to the latest v5, from JSdelivr : something like `boosted@5`. This way you'd stay up-to-date with npm for free.

Comment (julien-deramond) [reply .storybook/preview-head.html:3]: 😲 Seems logic but I hadn't thought about it. Thx @ffoodd!

Comment (julien-deramond) [reply .storybook/preview-head.html:3]: Hum, but it also means that we need to remove the `integrity` like that:
```html
<link href="https://cdn.jsdelivr.net/npm/boosted@5/dist/css/boosted.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdn.jsdelivr.net/npm/boosted@5/dist/js/boosted.bundle.min.js" crossorigin="anonymous"></script>
```
Given the fact that it is only used in the context of Storybook (and Zeroheight via an `<iframe>`) and not in a real application/website, I would say this security matter it is not that important. Not 100% sure about that. What do you think?

Comment (ffoodd) [reply .storybook/preview-head.html:3]: Agreed, as long as Boosted npm package depends on this repo I think it's sage 👌

**Comment by julien-deramond** [comment]
Still some bugs with some components but the full-automatic generation seems to work with https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/pull/565/commits/e9b4254c0fafc3a8631fc64c62ecab1f557e00c0.

For each PR, Netlify will generate the Storybook part.

For each release, the maintainer will have to run `npm run storybook-build` and copy the content of `site/storybook` in `gh-pages` branch.

**Comment by julien-deramond** [comment]
Still got some problems with Puppeteer even after https://github.com/Orange-OpenSource/Orange-Boosted-Bootstrap/pull/682
```
8:06:17 AM: creating /opt/build/repo/stories/auto/Navbar/Navbar_19.stories.js...
8:07:18 AM: { TimeoutError: Navigation timeout of 60000 ms exceeded
8:07:18 AM:     at Promise.then (/opt/build/repo/node_modules/puppeteer/lib/cjs/puppeteer/common/LifecycleWatcher.js:106:111) name: 'TimeoutError' }
```

**Comment by julien-deramond** [comment]
_This comment will be updated; it will help me to not to forget the issues_

## Review of components

### Accordion
- ✅ Rendering
- ✅ Manipulation

### Alerts
- ✅ Rendering
- ✅ Manipulation

### BackToTop
- ✅ Rendering
- ⚠️ Manipulation
  - Clicking on the buttons for each story opens a new page. Adding a `<a id="top">` doesn't seem to work but I need to dig into this problem a bit more. Even if we don't see the scroll, at least, we shouldn't change the current page.

### Badge
- ✅ Rendering
- ✅ Manipulation

### Breadcrumb
- ✅ Rendering
- ⚠️ Manipulation
  - Clicking on the link change the current page 

### ButtonGroup
- ✅ Rendering
- ⚠️ Manipulation
  - Clicking on the button or dropdown links change the current page 

### Buttons
- 🚫 Rendering
  - Icons are not displayed in local execution (`npm run storybook`)
- ⚠️ Manipulation 
  - Clicking on links change the current page

### Card
- 🚫 Rendering
  - "Image cap" text is not centered
  - For _Card 6_, the footer is not well displayed (too close to the border) but the same behavior is observed in the last example of https://boosted.orange.com/docs/5.0/components/card/#list-groups
  - For _Card 19_ the "Card image" text is not as big as in the documentation
  - Sometimes, the cards haven't the same height in _Card 27_ (for example) but the same behavior is observed in the documentation
- ⚠️ Manipulation 
  - Clicking on links change the current page

### Carousel
- 🚫 Rendering
  - The slides name are too small compared to the documentation
- ✅ Manipulation  

### CloseButton
- 🚫 Rendering
  - The last example should have a black background
-  ✅ Manipulation 

### Collapse
- ✅ Rendering
- ✅ Manipulation

### Dropdowns
TODO: review it

### ListGroup
TODO: review it

### Modal
TODO: review it

### Navbar
TODO: review it

### Navs
TODO: review it

### NavsTabs
TODO: review it

### OffCanvas
TODO: review it

### Pagination
TODO: review it

### Popovers
TODO: review it

### Progress
TODO: review it

### Scrollspy
TODO: review it

### Spinners
TODO: review it

### SteppedProcess
TODO: review it

### Toasts
TODO: review it

### Tooltips
TODO: review it

**Comment by ffoodd** [comment]
Regarding spinners, grow spinners were removed to match the brand.

Page changes are probably caused by Storybook handling for iframes, I guess ?

**Comment by julien-deramond** [comment]
> Regarding spinners, grow spinners were removed to match the brand.

Thanks for the history :) Comment updated.

> Page changes are probably caused by Storybook handling for iframes, I guess ?

Yes, I haven't studied it for the moment. I hope there is a way to bypass this behavior.

File: stories/create-stories-from-doc.js
Comment (julien-deramond) [inline stories/create-stories-from-doc.js:56]: This is my first idea to make components that use JavaScript work in Storybook: adding the specific JavaScript part at the end of the `.stories.js` file.
Another approach would be to add ALL the JavaScript specific parts within the global `preview-head.html` file; but it means to load unused things for nothing and also to have specific content separated from the automatic generator of stories.

Comment (julien-deramond) [reply stories/create-stories-from-doc.js:56]: I'm struggling to find a good architecture for all of this!
I'm wondering if I shouldn't use/embed _site/assets/application.js_ and the CSS in _site/assets/scss_: it would fix the JavaScript issues automatically (I think) and the specific rendering issues of the modals for example.
But if I do that, I'll need to add a `<div class="bd-example bd-example-*">` around the code in Storybook that will be displayed when the user will click on the "Code" tab...

**Comment by julien-deramond** [comment]
- [ ] Try to remove the titles
- [ ] What is "Nothing found"?

Capture d’écran 2021-06-23 à 14 33 14

**Comment by julien-deramond** [comment]
- [ ] Work on the ⚠️ message displayed while running `npm run storybook`:
```
(node:11682) DeprecationWarning: Default PostCSS plugins are deprecated. When switching to '@storybook/addon-postcss',
you will need to add your own plugins, such as 'postcss-flexbugs-fixes' and 'autoprefixer'.

See https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#deprecated-default-postcss-plugins for details.
(Use `node --trace-deprecation ...` to show where the warning was created)
```

**Comment by julien-deramond** [comment]
## Possible improvements after this 1st version

- Zeroheight will display all the "Buttons" examples in the "Develop" tab. But the problem is that some Boosted examples (from the documentation) doesn't exist in the UI designs. We would like to not display them in the Design System. At least two options to do that:
  - Not extracting in Storybook some example maybe by adding a specific class in the documentation to identify those examples
  - Not regrouping the components in packages (Buttons > Button 0 ; Buttons > Button 1 ; etc.) but it means that they should be added manually in the Design System. This causes a maintenance problem on the Design System side; and Storybook would be less useable without packages.

File: stories/create-stories-from-doc.js
Comment (julien-deramond) [reply stories/create-stories-from-doc.js:72]: Note: The designers embed the "Docs" part and the JavaScript is not executed there...

**Comment by julien-deramond** [comment]
Thank you @vprothais for the tip about `transformSource` in a3a39630af3b2b1de0072bebb6cc63f5b993db7c 👌

**Comment by julien-deramond** [comment]
Feedback from designers regarding the integration in ZH: is it possible to remove the margins?
Personal note: think to the use case of tooltips and popovers
Capture d’écran 2022-02-28 à 13 48 23

**Comment by julien-deramond** [comment]
Designer feedback: see if this blue link could be modified

Capture d’écran 2022-02-28 à 15 50 33

**Comment by julien-deramond** [comment]
There is a new issue because of the Tag component 😄 

Screenshot 2022-11-07 at 13 45 04

File: .storybook/storybook.scss
Comment (julien-deramond) [inline .storybook/storybook.scss:13]: Change with 0.8rem probably

File: .storybook/storybook.scss
Comment (julien-deramond) [inline .storybook/storybook.scss:5]: Find another value

File: stories/create-stories-from-doc.js
Comment (julien-deramond) [inline stories/create-stories-from-doc.js:119]: "boosted@5" -> "boosted"

File: package.json
Comment (louismaximepiton) [inline package.json:100]: Shouldn't we add this command to the release one ?

**Review by louismaximepiton — COMMENTED** [review]
Looks like the Hash of `boosted.css` and `boosted.bundle.js` are the same between main branch and this one. Do I need to chack all of them ?
We don't have any images like logos, or icons/images in localhost.

The changes seem to be great, but I have several questions before merging.

File: config.yml
Comment (julien-deramond) [reply config.yml:55]: It is useful to have access to `/storybook` in the URL (e.g. http://localhost:9001/storybook in local mode)

File: stories/create-stories-from-doc.js
Comment (julien-deramond) [reply stories/create-stories-from-doc.js:84]: It doesn't work on my side. I have the same error as you when we talked about it this morning:

```
"file:/Users/ju/Orange-Boosted-Bootstrap/_site/storybook/index.html":1.38-1.60: info warning: Self-closing tag syntax in text/html documents is widely discouraged; it's unnecessary and interacts badly with other HTML features (e.g., unquoted attribute values). If you're using a tool that injects self-closing tag syntax into all void elements, without any option to prevent it from doing so, then consider switching to a different tool.
"file:/Users/ju/Orange-Boosted-Bootstrap/_site/storybook/index.html":1.87-1.154: info warning: Self-closing tag syntax in text/html documents is widely discouraged; it's unnecessary and interacts badly with other HTML features (e.g., unquoted attribute values). If you're using a tool that injects self-closing tag syntax into all void elements, without any option to prevent it from doing so, then consider switching to a different tool.
ERROR: "docs-lint" exited with 1.
```

File: stories/create-stories-from-doc.js
Comment (julien-deramond) [reply stories/create-stories-from-doc.js:41]: Yes I'm going to add it in the issue that will be created after the merge. Agreed that we need to make things automatic.

File: .storybook/preview-head.html
Comment (julien-deramond) [reply .storybook/preview-head.html:7]: IDK this is the copyright used everywhere

File: .storybook/main.js
Comment (julien-deramond) [reply .storybook/main.js:16]: Nope removed it

File: .storybook/main.js
Comment (julien-deramond) [reply .storybook/main.js]: Don't think it is really used so I removed it

File: .storybook/main.js
Comment (julien-deramond) [reply .storybook/main.js:10]: Probably not but too much to investigate right now. I'll put it in the issue created after the merge.

File: .storybook/main.js
Comment (julien-deramond) [reply .storybook/main.js:15]: Yep otherwise

```
ERROR in ./.storybook/storybook.scss 1:12
Module parse failed: Unexpected token (1:12)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> pre.prismjs {
|   color: #fff;
|   font-size: .875rem;
 @ ./.storybook/preview.js 4:0-26
 @ ./.storybook/preview.js-generated-config-entry.js
 @ multi ./node_modules/@storybook/core-client/dist/esm/globals/polyfills.js ./node_modules/@storybook/core-client/dist/esm/globals/globals.js (webpack)-hot-middleware/client.js?reload=true&quiet=false&noInfo=undefined ./storybook-init-framework-entry.js ./node_modules/@storybook/html/dist/esm/client/preview/config-generated-config-entry.js ./node_modules/@storybook/html/dist/esm/client/docs/config-generated-config-entry.js ./node_modules/@storybook/addon-a11y/preview.js-generated-config-entry.js ./node_modules/@storybook/addon-docs/preview.js-generated-config-entry.js ./node_modules/@storybook/addon-actions/preview.js-generated-config-entry.js ./node_modules/@storybook/addon-backgrounds/preview.js-generated-config-entry.js ./node_modules/@storybook/addon-outline/preview.js-generated-config-entry.js ./.storybook/preview.js-generated-config-entry.js ./generated-stories-entry.js
```

File: package.json
Comment (julien-deramond) [reply package.json:100]: I let's try it

**Review by louismaximepiton — APPROVED** [review]
LGTM, let's try it

**Pull request merged**
